jni/src/main/java/com/horizen/commitmenttree/CommitmentTree.java

```diff
@@ -0,0 +1,201 @@
+                certVerificationKey, btrVerificationKeyOpt.orElse(null), cswVerificationKeyOpt.orElse(null),
+                txHash, outIdx);
+    }
+    
+    private native boolean nativeAddFwt(byte[] scId, long amount, byte[] pubKey, byte[] txHash, int outIdx);
+    
+    public boolean addFwt(byte[] scId, long amount, byte[] pubKey, byte[] txHash, int outIdx) {
+        if (commitmentTreePointer == 0)
+            throw new IllegalStateException("CommitmentTree instance was freed.");
+        return nativeAddFwt(scId, amount, pubKey, txHash, outIdx);
+    }
+
+    private native boolean nativeAddBtr(byte[] scId, long scFee, byte[] pubKeyHash,
+                                               byte[] scRequestData, byte[] txHash, int outIdx);
+    
+    public boolean addBtr(byte[] scId, long scFee, byte[] pubKeyHash, byte[] scRequestData, byte[] txHash, int outIdx) {
+        if (commitmentTreePointer == 0)
+            throw new IllegalStateException("CommitmentTree instance was freed.");
+        return nativeAddBtr(scId, scFee, pubKeyHash, scRequestData, txHash, outIdx);
+    }
+    
+    private native boolean nativeAddCert(byte[] scId, int epochNumber, long quality,
+                                                byte[] certDataHash, BackwardTransfer[] btList,
+                                                byte[] customFieldsMerkleRoot,
+                                                byte[] endCumulativeScTxCommitmentTreeRoot);
+
+    public boolean addCert(byte[] scId, int epochNumber, long quality, byte[] certDataHash, BackwardTransfer[] btList,
+                           byte[] customFieldsMerkleRoot, byte[] endCumulativeScTxCommitmentTreeRoot) {
+        if (commitmentTreePointer == 0)
+            throw new IllegalStateException("CommitmentTree instance was freed.");
+        return nativeAddCert(scId, epochNumber, quality, certDataHash, btList,
+                customFieldsMerkleRoot, endCumulativeScTxCommitmentTreeRoot);
+    }
+
+    public native boolean nativeAddCertLeaf(byte[] scId, byte[] leaf);
+
+    public boolean addCertLeaf(byte[] scId, byte[] leaf) {
+        if (commitmentTreePointer == 0)
+            throw new IllegalStateException("CommitmentTree instance was freed.");
+        return nativeAddCertLeaf(scId, leaf);
+    }
+    
+    private native boolean nativeAddCsw(byte[] scId, long amount, byte[] nullifier,
+                                               byte[] pubKeyHash, byte[] activeCertDataHash);
+    
+    public boolean addCsw(byte[] scId, long amount, byte[] nullifier, byte[] pubKeyHash, byte[] activeCertDataHash) {
+        if (commitmentTreePointer == 0)
+            throw new IllegalStateException("CommitmentTree instance was freed.");
+        return nativeAddCsw(scId, amount, nullifier, pubKeyHash, activeCertDataHash);
+    }
+
+
+    private native Optional<FieldElement[]>  nativeGetCrtLeaves(byte[] scId);
+    public Optional<List<FieldElement>> getCrtLeaves(byte[] scId) {
```

Comment:
Looks like this method is not covered with tests in CommitmentTreeTest

Reply:
Done.